### ISSUE TITLE
Heatmap bins do not line up with annotations

### ISSUE THREAD

**Issue by nollikai** [higlass/higlass#1051]
_Commit_: 

#### Steps to reproduce
1. Go to a HiGlass example page: http://higlass.io/app/?config=OPPFtzbuR76vtUIshTEOGA
2. Set zoom limit to 5kb
3. Using hg38, zoom to region "chr2:51,793-82,274 & chr2:51,957-67,232 [offset 0,0:0,0]"

#### Observed behavior
- Heatmap 5kb bins do not line up with annotated 5kb bins (shown in attached image as dashed lines)

<img width="993" alt="higlass_shifted_annotation" src="https://user-images.githubusercontent.com/4184083/131892886-1e6d3b62-5113-48fd-b922-a4e064f13350.png">

#### Expected behavior
- Heatmap bins should line up with annotated bins

**Comment by pkerpedjiev** [comment]
Surprisingly, this is actually not a bug and is expected. We should write a blog post about it or document it since it's come up fairly often.

The core issue is that we line chromosomes up from end to end without regard to bin size. That means that bins at the end of a chromosome are almost always truncated. So what do we do?

Well, we always display each bin in full and pull bins based on where they start. So we would get something like the following:

```
          chr1             chr2     chr3
|--------------|---------|-----|            chromosome boundaries
|----|----|----|----|----|----|----|         displayed bins
b1.1  b1.2 b1.3 b2.1 b2.2 b3.1 b3.2  
|----|----|----||----|----||----|-|           actual bins
b1.1  b1.2 b1.3 b2.1. b2.2 b3.1 b3.2
                    b1.4          b2.3
```

The bins are slightly offset from their boundaries but always contain their starting point. This only a display artifact, however. The stored bins do actually align with the data boundaries.

I hope this makes some sense.

**Comment by golobor** [comment]
oh, i actually did not know that! Sorry for bugging you, I am not sure if I fully understood - do these mismatches add up with each chromosome or do they get "reset" at chromosomal boundaries? If latter, does it mean that the bins of chromosomes that appear later in the genome (say, chr22 in hg) are generally less aligned than those of "earlier" chromosomes? If yes, could this result in misalignments that are greater than a bin size?..

**Comment by alexpreynolds** [comment]
This comment might be useful, maybe not. I'll mention it in case it is
helpful.

The translation between genomic coordinates and absolute units along a line
can result in misalignments further along the chromosomal ordering, past
the first chromosome, because genomic coordinates of annotations or other
data past the first chromosome do not generally translate to neat absolute
units.

One solution I have used is to create a synthetic assembly for the
chromosomes I need, for the data I'm working with.

For example, we have one site which uses HiGlass to show data that are
represented by quantitative signal at 75 and 200nt units, over chromosomes
chr1 through chrY.

To resolve alignment issues between different tracks, we create a new
"assembly" that terminates each chromosome at their least-common-multiple
of 75 and 200, which is 600.

Say we are working with hg38. We get chromSizes from UCSC or elsewhere. The
length of chr1 is 248,956,422 bases. We bring this up to the next multiple
of 600, which is 248,956,800. The length of chr2 is 242,193,529 bases,
which is brought to 242,193,600. And so on for the remaining chromosomes.

When we make all the tracks we need with clodius etc. we use this synthetic
chromSizes file when generating tiles. The faked sizes of chromosomes make
it so that the ends of annotations will always fall on an integer-based
partitioning of the absolute coordinate system that HiGlass uses.

For us, this has resolved alignment issues between tracks. However, one
downside is that it locks us into units that fit the LCM of those tracks we
are currently showing. If we needed to show new tracks down the road that
do not fit the LCM, we may need to reencode all tracks to line up with a
new LCM. Another downside is having a little extra empty data at the end of
each chromosome. But so far this is working for us.

-Alex

**Comment by Yaoyx** [comment]
> Surprisingly, this is actually not a bug and is expected. We should write a blog post about it or document it since it's come up fairly often.
> 
> The core issue is that we line chromosomes up from end to end without regard to bin size. That means that bins at the end of a chromosome are almost always truncated. So what do we do?
> 
> Well, we always display each bin in full and pull bins based on where they start.


Hi @pkerpedjiev, 

we may be confused about your reply, but it does not seem that each bin is displayed in full for HiC data. Below is a minimal example with toy data that shows the second bin on chr1, which is less than a full binsize yet has nonzero pixels, is not displayed. Moreover, the genomic position coordinate above the map is out of sync with what is displayed. This also presents issues when co-visualizing other data, like bigwigs.

```python
import pandas as pd
import cooler
# Generate test matrix with weights
bins=pd.DataFrame(
    [["chr1", 0, 5, 1], 
        ["chr1", 5, 8, 1], 
        ["chrX", 0, 5, 1],
        ["chrX", 5, 10, 1]],
columns=["chrom", "start", "end", "weight"],
)

pixels = pd.DataFrame(
    [[0, 0, 6], [0, 1, 10], [0, 2, 2], [0, 3, 1],
     [2, 3, 3],
     [3, 3, 0]
     ], 
    columns=["bin1_id", "bin2_id", "count"]
)

clr_file = "/home1/yxiao977/sc1/train_akita/test_data/test_coverage.cool"
cooler.create_cooler(clr_file, bins, pixels)
```

``` 
! cooler zoomify '/home1/yxiao977/sc1/train_akita/test_data/test_coverage.cool' -o '/home1/yxiao977/sc1/train_akita/test_data/Multires_test_coverage.mcool' -r 5,10 --balance```
```

```python
import higlass as hg
hg.server.enable_proxy()
ts7 = hg.cooler('/home1/yxiao977/sc1/train_akita/test_data/Multires_test_coverage.mcool')
hg.view(ts7.track("heatmap"), ts7.track("chromosome-labels"))
```

<img width="1828" alt="image" src="https://github.com/higlass/higlass/assets/37345096/50d315c2-e804-46f0-85d7-79a7c318189a">

It would be really great to have this documented & explain how we can co-visualize bigwigs without an offset between the HiC map and the track.

Thank you !
Yao & @gfudenberg

**Comment by pkerpedjiev** [comment]
Sorry for the delayed reply. Your heatmap does appear to be missing a bin but I still *think* that that's expected. I'll explain with code rather than diagrams. If you'd like to follow with context, I'm starting here: https://github.com/higlass/clodius/blob/develop/clodius/tiles/cooler.py#L322

The first axiom is that higlass operates on a binned genome rather than chromosomes. This was a design decision made early on that lets us rasterize tiles across chromosome boundaries. What is a genome? Well it's just a concatenation of chromosomes. Normally this is defined by a set of chromosome sizes. In your case we can infer it from the bins:

```
chr1 8
chrX 10
```

You can see that your genome has a length of 18 (8 + 10). Now we can remove the chromosome labels from your bins and write them as genome coordinates:

```
 "chr1", 0, 5 -> 0, 5
  "chr1", 5, 8 -> 5, 8
  "chrX", 0, 5 -> 8, 13
  "chrX", 5, 10 -> 13, 18
```

Indeed, when we filter out the data to display for a tile, we refer `genome_start`:

```
            df = data[data["genome_start1"] >= start1]
            df = df[df["genome_start1"] < end1]

            df = df[df["genome_start2"] >= start2]
            df = df[df["genome_start2"] < end2]
```


Now if we assume a base resolution of 5, then the genome will be display bins at positions: 0, 5, 10, 15 (assuming 0-based coordinates).  You can immediately see that after the first chromosome, the genome bins don't overlap exactly with the chromosome bins.  This is a known effect and it's tradeoff we made early on under the assumption that the vast majority of the time viewers would not be looking at a high enough resolution for the bin boundaries to make a difference.

So what's happening with your example? Well we can calculate the genome bins that are being assigned to each pixel:

```
            j = ((df["genome_start1"].values - start1) // binsize).astype(int)
            i = ((df["genome_start2"].values - start2) // binsize).astype(int)

           print("df", df)
           print("j", j)
           print("i", i)
```

Which yields:

```
df    genome_start1  genome_start2  balanced
0              0              0         6
1              0              5        10
2              0              8         2
3              0             13         1
4              8             13         3
5             13             13         0
j [0 0 0 0 1 2]
i [0 1 1 2 2 2]
```

We we can see that pixels 1 and 2 are assigned to the same bin (0,1). The y-bin for pixel 2 is being eaten up by the y-bin for pixel 1. Does that mean that bigwigs and coolers will be out of sync? It depends. We employ (I believe) the same strategy for pulling bigwigs values so as long as they contain the same resolutions then they should be aligned. If they don't contain the same resolutions then there's a good chance they wouldn't be aligned even if higlass didn't use genome wide bins.

Does this chromosome boundary effect accumulate across chromosomes leading to larger offset errors? It does not. Because we interpolate chromosome-based bins from genome-based coordinates, we always pull whatever chromosome bin start lies within the genome bin. So in the example above, this led to a bin not appearing. The opposite effect can also happen where a smaller bin appears larger. Imagine if we added one more pixel:

```
 "chr1", 0, 5 -> 0, 5
  "chr1", 5, 8 -> 5, 8
  "chrX", 0, 5 -> 8, 13
  "chrX", 5, 10 -> 13, 18
  "chrX", 10, 12 -> 18, 20
```

Now we would end up with the small 2nt pixel appearing larger:

<img width="458" alt="image" src="https://github.com/higlass/higlass/assets/2143629/a9c36c50-6e1f-4419-aba2-1801b2d0824f">

So in summary, individual pixels are not nt-perfect past the first chromosome. They are, however, guaranteed to contain their start value.

**Comment by Yaoyx** [comment]
> We we can see that pixels 1 and 2 are assigned to the same bin (0,1). The y-bin for pixel 2 is being eaten up by the y-bin for pixel 1. Does that mean that bigwigs and coolers will be out of sync? It depends. We employ (I believe) the same strategy for pulling bigwigs values so as long as they contain the same resolutions then they should be aligned. If they don't contain the same resolutions then there's a good chance they wouldn't be aligned even if higlass didn't use genome wide bins.

Hi @pkerpedjiev,

I generated a toy bigwig data based on the previous example. The resolution of bigwig data is set to 5, but it seems that bigwig track does not use the same strategy as heatmap. The bin2 column partially corresponds to 0.6 (cis/total ratio of bin2) on bigwig track, but if it follows the same strategy, this column should only correspond to 0.6.

```python
# toy bigwig file 
import cooler
import pandas as pd

clr = cooler.Cooler("/home1/yxiao977/sc1/train_akita/test_data/test_coverage.cool")
bins_table = clr.bins()[:]
cis_total_cov_df = pd.DataFrame([0.2, 0.4, 0.6, 0.8], columns=['cis-total-ratio'])
bins_table = pd.concat([bins_table, cis_total_cov_df], axis=1)

import pyBigWig as pbw

bw = pbw.open("/home1/yxiao977/sc1/train_akita/test_data/test_coverage.bw", 'w')
chromsizes = list(zip(clr.chromnames, clr.chromsizes))
bw.addHeader(chromsizes)

chroms, starts, ends, weights, ratios = list(zip(*bins_table.values))

bw.addEntries(list(chroms), list(starts), ends=list(ends), values=list(ratios))
bw.close()
```
```python
import higlass as hg
hg.server.enable_proxy()

ts1 = hg.cooler('/home1/yxiao977/sc1/train_akita/test_data/Multires_test_coverage.mcool')
ts2 = hg.bigwig('/home1/yxiao977/sc1/train_akita/test_data/test_coverage.bw')
hg.view(ts1.track('heatmap'), ts1.track('chromosome-labels'), ts2.track('horizontal-bar'))
```
![image](https://github.com/higlass/higlass/assets/37345096/c09949fb-0379-4ad1-9134-1b5db5999ae4)

**Comment by pkerpedjiev** [comment]
So my "I believe" statement turned out to be false. I should have double checked before lazily claiming that. There were in fact two things that I incorrectly assumed:

1. That bigwig tile fetching is done as with cooler tile fetching. This is not the case. With cooler files we retrieve multi-chromosome "tiles" as rasterized grids. That's why, as in the example above, some bins may be skipped. Bigwig tiles are fetched per chromosome. So if we have a tile that spans multiple chromosomes, we fetch the constituent portion from each chromosome. You can see that happen here: https://github.com/higlass/clodius/blob/develop/clodius/tiles/bigwig.py#L162

This is the "better" way to do this because we can dynamically rearrange chromosomes with bigwig files. Unfortunately, cooler files have O(n ** 2) chromosome interactions which would (I think) put a hefty computational strain on the server to pull out piecewise.

2. The resolution of the bigwig file is independent of the bins passed in. In the cooler file, the highest resolution was 5. We can't zoom in and see more resolution because there simply isn't any there (except for, of course, the trailing bins). Bigwig, on the other hand, will happily give 1bp resolution even if it was created with 5bp bins. That lets us zoom in beyond the 'base' size and see that trailing bin.

So where do we go from here? I'll ask you: how important is it to have perfect alignment between bigwig and cooler bins?

A fairly simple solution would be to apply point #2 to cooler files and allow fetching of rasterized tiles below the minimum resolution. We would still return an evenly sized grid of data to the client but because the resolution of the grid would be higher than the base bin size, the trailing bins would be visible.

Actually, as I write this I realize that even if it's not something that's super important it's the correct way to display this data. I'll try to get it implemented in the next week or two.

Thanks a ton for bringing this back up and for the detailed code snippets and examples! It really made it clear what was happening and how it could be fixed.
